contracts/src/v0.8/dev/ocr2dr/OCR2DRRegistry.sol

```diff
@@ -729,6 +736,33 @@ contract OCR2DRRegistry is ConfirmedOwner, OCR2DRRegistryInterface, ERC677Receiv
     return false;
   }
 
+  /**
+   * @notice Time out all expired requests: unlocks funds and removes the ability for the request to be fulfilled
+   * @param subscriptionId - ID of the subscription
+   */
+
+  function timeoutRequests(uint64 subscriptionId) external onlySubOwner(subscriptionId) {
+    address[] memory consumers = s_subscriptionConfigs[subscriptionId].consumers;
+    address[] memory authorizedSendersList = getAuthorizedSenders();
+    for (uint256 i = 0; i < consumers.length; i++) {
+      for (uint256 j = 0; j < authorizedSendersList.length; j++) {
+        bytes32 requestId = computeRequestId(
```

Comment:
This seems very inefficient.  Can we store in-flight requestIds for each subscription in an array instead of needing to iterate thru the entire list of authorized senders and computing the requestId for every single possible requestId?

ie: we create a new map `mapping(subscriptionId => inFlightRequestIds[])`.  Then, we only need to check if any of the inFlightRequestIds for are expired, which should be much more efficient.

Reply:
Actually, why do we need to iterate thru the list of authorized senders at all?  Can't we just use the subscription owners request?  Each individual subscription owner should be responsible for timing out their own unfulfilled requests, not paying for gas to time out everyone's unfulfilled requests.

Reply:
(If we are able to just check for timed out requests _only_ for the subscription owner who is calling this contract, ignore my first comment about creating an array)

Reply:
My suggested changes:

Add a `view` function: `getRequestsToTimeout(uint64 subscriptionId) returns (bytes32[] requestIdsToTimeout)`

This function should iterate thru all `consumers` for the input `subscriptionId`, generate the requestIds and check if those requestIds need to be timed out.  Then, return a list of all requestIds which need to be timed out.


Modify `timeoutRequests(bytes32[] requestIdsToTimeout)` to iterate through the input `requestIdsToTimeout`, check that they are actually timed out, unlock funds, and delete the requestId commitment.

Finally, I don't see any reason to have the `onlySubOwner` modifier.  IMO, _anyone_ should be able to timeout requests and unlock funds.  That gives us the flexibility for CLL to timeout requests instead of forcing users to do (it if for some reason we want to do that for example, if we have some outage that is our fault, idk)

Reply:
> I don't see any reason to have the onlySubOwner modifier. IMO, anyone should be able to timeout requests and unblock funds.

Not sure that I agree with this. I could see opening it to all sub consumers, but opening it to _anyone_ just seems like it could leave a hole for abuse. If I make a request, it goes over the allowed time, but I don't mind if it's stale - I wouldn't want some random person to time out my request.

Reply:
> Actually, why do we need to iterate thru the entire list of authorized senders at all? Can't we just use the subscription owner's address?
Not to time out everything on the sub. Each sub consumer has its own nonce that is use to generate the request ID.

> Each individual subscription owner should be responsible for timing out their own unfulfilled requests, not paying for gas to time out everyone's unfulfilled requests.

It doesn't pay for gas for anyone not on the sub

Reply:
I'm realizing that we will run into the issue of "only the latest nonce is persisted" thing here.. so definitely should change to timing out by requestId.

`getRequestsToTimeout` still has the nonce issue. May be better to handle the finding of stale requestIds completely off-chain.

Reply:
> This seems very inefficient. Can we store in-flight requestIds for each subscription in an array instead of needing to iterate thru the entire list of authorized senders and computing the requestId for every single possible requestId?

Storage is the most expensive part of blockchains, so if the concern is gas cost, then storing will cost more than iterating.

Reply:
> > I don't see any reason to have the onlySubOwner modifier. IMO, anyone should be able to timeout requests and unblock funds.
> 
> Not sure that I agree with this. I could see opening it to all sub consumers, but opening it to _anyone_ just seems like it could leave a hole for abuse. If I make a request, it goes over the allowed time, but I don't mind if it's stale - I wouldn't want some random person to time out my request.

My preference would be that there is no such thing as a stale request.  It either fulfills within the given time or fails.  While I would be curious to hear other opinions, I like this functionality because it leaves no question as to whether a request _might_ be fulfilled.  As a blockchain developer, I am totally averse to ambiguity.

Reply:
> I'm realizing that we will run into the issue of "only the latest nonce is persisted" thing here.. so definitely should change to timing out by requestId.
> 
> `getRequestsToTimeout` still has the nonce issue. May be better to handle the finding of stale requestIds completely off-chain.

I am slightly concerned about this.  Before I would approve of this approach, I would want us to have a clear way of generating a list of timed out requests for a particular sub.  Simply checking for all past events that contain a requestId is a huge pain.  If I am a dev who isn't keeping track of all of their requestIds and I don't want to have to check all past logs, how do I get my requestIds?

Reply:
> > This seems very inefficient. Can we store in-flight requestIds for each subscription in an array instead of needing to iterate thru the entire list of authorized senders and computing the requestId for every single possible requestId?
> 
> Storage is the most expensive part of blockchains, so if the concern is gas cost, then storing will cost more than iterating.

Yeah, that is true.  But at a certain point if the list gets large enough, that may no longer become the case.  IDK, anyway it is a mute point now since we are gonna implement cancellation directly by requestId anyway.

Reply:
Note: I am approving this right now.  However, we need to have a discussion about a user will generate a list of requestIds given they are not written in the contract.

Some suggestions so far:
- a `view` function: `getPendingRequestIds(uint64 subscriptionId, address consumer) returns (bytes32[] requestIds)`
  - This function would iterate from the current consumer's nonce _n_ to 0.  It generates each possible requestId and checks if that requestId has a corresponding commitment in the commitments mapping.
- Completely off-chain approach
  - Create tooling to interact with Atlas or The Graph such that a user can generate a list of all past requestIds for their subscription using past emitted events.